File: python/3.py
with open("../3.txt") as f:
    banks = [x.strip() for x in f]


def get_joltage(bank: str, cells: int) -> int:
    length = len(bank)
    joltage = []
    index = 0
    for i in range(cells):
        joltage.append((largest := max(bank[index : length - (cells - 1 - i)])))
        index = bank.index(largest, index) + 1
    return int("".join(joltage))


print(sum(get_joltage(bank, 2) for bank in banks))

print(sum(get_joltage(bank, 12) for bank in banks))

File: rust/src/bin/3.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::str::FromStr;

fn main() {
    let banks: Vec<_> = BufReader::new(File::open("../3.txt").unwrap())
        .lines()
        .map(|l| l.unwrap())
        .collect();

    println!(
        "{}",
        banks.iter().map(|bank| get_joltage(bank, 2)).sum::<usize>()
    );

    println!(
        "{}",
        banks
            .iter()
            .map(|bank| get_joltage(bank, 12))
            .sum::<usize>()
    );
}

fn get_joltage(bank: &str, cells: usize) -> usize {
    let length = bank.len();
    let mut joltage = String::with_capacity(cells);
    let mut index = 0;
    for i in 0..cells {
        let largest = bank[index..length - (cells - 1 - i)].chars().max().unwrap();
        index += bank[index..].find(largest).unwrap() + 1;
        joltage.push(largest);
    }
    usize::from_str(&joltage).unwrap()
}
